### PR TITLE
thanos: support arbitrary object storage configuration

### LINKED ISSUE #2133: [thanos] Support for Azure Storage
The current documentation indicates that it's possible to use S3 and GCS but Azure is not available. Azure support was implemented relatively recently and it would be great to get this functionality into prometheus-operator

### LINKED ISSUE #2225: Can't set objstore option for thanos.
**What did you do?**
Thanos latest version supports objstore.config options instead of s3.xxx (these are removed)
So when I use set thanos v0.2.0, storage configuration is not applied properly.

--objstore.config-file=/etc/thanos/thanos-s3.yaml

https://github.com/improbable-eng/thanos/blob/master/cmd/thanos/flags.go

```go
func regCommonObjStoreFlags(cmd *kingpin.CmdClause, suffix string) *pathOrContent {
	fileFlagName := fmt.Sprintf("objstore%s.config-file", suffix)
	bucketConfFile := cmd.Flag(fileFlagName, fmt.Sprintf("Path to YAML file that contains object store%s configuration.", suffix)).
		PlaceHolder("<bucket.config-yaml-path>").String()

	bucketConf := cmd.Flag(fmt.Sprintf("objstore%s.config", suffix), fmt.Sprintf("Alternative to '%s' flag. Object store%s configuration in YAML.", fileFlagName, suffix)).
		PlaceHolder("<bucket.config-yaml>").String()

	return &pathOrContent{
		name: fmt.Sprintf("objstore%s.config", suffix),

		path:    bucketConfFile,
		content: bucketConf,
	}
}
```

I can't find any related code in prometheus-operator.

**What did you expect to see?**
I want to add objstore.config and objstore.config.file options.

**What did you see instead? Under which circumstances?**

**Environment**

* Prometheus Operator version:

	insert image tag or Git SHA here

* Kubernetes version information:

	insert output of `kubectl version` here

* Kubernetes cluster kind: 

	insert how you created your cluster: kops, bootkube, tectonic-installer, etc.

* Manifests:

```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
insert Prometheus Operator logs relevant to the issue here
```

### PR DESCRIPTION
**Pull request by abursavich**
[Thanos v0.2.0](https://github.com/improbable-eng/thanos/releases/tag/v0.2.0) introduced breaking changes that moved object storage configuration from a large set of flag and environment variables for each possible provider to one of two flags `--objstore.config` or `--objstore.config-file`.

This change updates prometheus-operator to support newer versions of thanos and decouples  prometheus-operator from the details of thanos object storage configuration.  Object storage configuration is stored in a user-created secret, about which prometheus-operator needs only to know the name.  Thanos is free to introduce support for arbitrary backends without updates to prometheus-operator.

Details:
- Adds a deprecation notice to GCS (ThanosGCSSpec) and S3 (ThanosS3Spec) in ThanosSpec.
- Adds an ObjectStorageConfig SecretKeySelector in ThanosSpec.
- When specified, adds the ObjectStorageConfig secret as an environment variable and adds the environment variable as the `--objstore.config` flag value.
- Bumps the default thanos version to v0.2.1.
- Regenerates code/docs with `make generate-in-docker`.

Fixes #2133 and fixes #2225

### CODE REVIEW

**Comment by abursavich** [comment]
Yeah, that's fine. I can take care of the rebase real quick.

**Comment by metalmatze** [comment]
I would like to rename the `ObjectStorage` to `ObjectStorageSecret` or `ObjectStorageConfig` to make it more obvious what type we're referencing, as we don't directly reference an object storage.
Any objections?

**Comment by abursavich** [comment]
I tried to make it similar to existing names: GCS/S3 fields in ThanosSpec don't have a Config suffix; Credentials in ThanosGCSSpec and AccessKey/SecretKey in ThanosS3Spec don't have a Secret suffix.

That said, I really don't have strong opinions about naming here.

**Comment by metalmatze** [comment]
Yep, I also compared to already given fields. The difference with those GCS/S3 fields is that those have been flags passed to the sidecar. I think would go with `ObjectStorageConfig` and the config is a secret. :slightly_smiling_face:

**Comment by abursavich** [comment]
Done.

**Comment by metalmatze** [comment]
Ah, oh. Yeah, I would have pushed that with my test, but I can rebase as well. Thanks :smile:

**Comment by metalmatze** [comment]
Alright. I've tested this on minikube and it works.

Tested with 
`prometheus.yaml`:
```diff
...
+  thanos:
+    peers: thanos-peers.monitoring.svc:10900
+    objectStorageConfig:
+      key: thanos.yaml
+      name: thanos-objstore-config
```

and the config being:
`thanos-objstore-config.yaml`:
```yaml
type: s3
config:
  bucket: testing
  endpoint: ams3.digitaloceanspaces.com
  access_key: sdfasdf
  secret_key: asdfsdfsad

# Use this command to generate a secret with this file
# kubectl -n monitoring create secret generic thanos-objstore-config --from-file=thanos.yaml=thanos-objstore-config.yaml
```

Tomorrow I'll replace this with my already running thanos setup and do some more long running tests. We should also update the docs after merging this PR.

File: pkg/prometheus/statefulset.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ import (
 const (
 	governingServiceName     = "prometheus-operated"
 	DefaultPrometheusVersion = "v2.5.0"
-	DefaultThanosVersion     = "v0.1.0"
+	DefaultThanosVersion     = "v0.2.1"
```
Comment (abursavich) [inline pkg/prometheus/statefulset.go]: It looks like there was a change in gofmt and the tests don't like your formatting of this file. For some reason (rebase?), the changes don't show up in the diff here.

Your formatting is in the v1.10 gofmt style, but the tests want the v1.11 gofmt style.

https://github.com/golang/go/issues/26098

**Comment by abursavich** [comment]
I looks like I broke the tests by using v1.10 gofmt and @metalmatze fixed it by using v1.11 gofmt, but for some reason Travis hasn't rerun the tests.

**Comment by metalmatze** [comment]
Strange, I only have Go 1.11 on my system. Seems to be good now after rebasing and pushing again.
These changes are running on my personal cluster without any problems. I'm going to merge as soon as a block got uploaded to object storage.

**Comment by metalmatze** [comment]
Works like a charm. Thanks everyone who contributed also in the other PR.

**Comment by szymonpk** [comment]
Thanks for working on this! Any chance for cutting the release anytime soon? It is very welcomed change ;-)

**Comment by brancz** [comment]
We just cut a release but we cut releases relatively frequently so you shouldn't need to wait too long :slightly_smiling_face:

**Comment by metalmatze** [comment]
There are a few things I want to address in the next days, however mostly for kube-prometheus before announcing a new Prometheus Operator release with Thanos v0.2 support.

* [ ] Add StatefulSet for Thanos store #2039 
* [ ] Add compactor to mixin (kube-prometheus-thanos.libsonnet)
* [ ] Add ServiceMonitors for all Thanos components to mixin
* [ ] Update docs reflecting the last changes

If you want to help with something let me know.

**Comment by brancz** [comment]
@kkc if you can create a pull request to add some docs on that, that would be fantastic! :slightly_smiling_face:

**Comment by boxuan666** [comment]
@metalmatze Hi seem like `ObjectStorageConfig ` is not working in Promethe-operator v0.33.0.

**Comment by boxuan666** [comment]
@abursavich

**Comment by abursavich** [comment]
@boxuan666, I don't see anything wrong with the thanos config snippet you posted, but I'm unable to reproduce your issue.

**Comment by boxuan666** [comment]
yep... how can I solve this problem? what advice do you have?

**Comment by abursavich** [comment]
moving conversation to https://github.com/coreos/prometheus-operator/issues/2869

**Pull request merged**